File: src/....JegasCRM-Backup/webshare/iwf/php/iwf.php
<?
class IWFAction {

	var $type;
	var $target;
	var $code;
	var $msg;
	var $content;
	var $emitted;

	function IWFAction($type, $target, $content){
		$this->type = $type;
		$this->target = $target;
		$this->emitted = false;
		$this->content = $content;
	}

	function error($code, $msg){
		$this->code = $code;
		$this->msg = $msg;
		$this->content = '';
		return $this;
	}

	function _echoStart($prettyprint){
		if ($prettyprint){
			echo "\n\t";
		}
		echo "<action type='$this->type'";
		if ($this->target){
			echo " target='$this->target'";
		}
		if ($this->errorCode){
			echo " errorCode='$this->code'";
		}
		if ($this->message){
			echo " message='$this->message'";
		}
		echo '>';
		if ($prettyprint){
			echo "\n\t\t";
		}
		// once we echo any part of this object, we must not re-echo it,
		// so mark this object as emitted.
		$this->emitted = true;
		return $this;
	}

	function _echoEnd($prettyprint){
		if ($prettyprint){
			echo "\n\t";
		}
		echo "</action>";
		return $this;
	}

	function emit($echoEnd, $prettyprint){
		if (!$this->emitted){
			$this->_echoStart($prettyprint);
			echo $this->content;
		}
		if ($echoEnd){
			$this->_echoEnd($prettyprint);
		}
		return $this;
	}
}

class IWFController {

	var $actions;
	var $currentAction;
	var $currentlyEmitting;
	var $startedEmitting;
	var $debugging;
	var $prettyprint;
	var $iframeName;

	function IWFController(){
		if ($iwf){
			alert("Do not create new IWFController objects.\nThe singleton stored in $iwf should be used instead.");
			return;
		}
		$this->actions = array();
		$this->currentAction = null;
		$this->currentlyEmitting = '';
		$this->startedEmitting = false;
		$this->debugging = false;
		$this->prettyprint = false;
		$this->iframeName = '';
	}

	function debug($on){
		$this->debugging = $on;
		if ($on){
			$this->prettyprint = true;
		}
	}

	function pretty($on){
		$this->prettyprint = $on;
	}

	function iframe($name){
		$this->iframeName = $name;
	}

	function startAction($type, $target){
		// ensure we've emitted everything up until this point...
		$this->_emit(false);
		$this->currentAction = new IWFAction($type, $target, '');
		$this->currentAction->emit(false, $this->prettyprint);
		$this->currenltyEmitting = $type;
		return $this->currentAction;
	}

	function endAction(){
		if ($this->currentAction){
			$this->currentAction->emit(true, $this->prettyprint);
			$this->currentlyEmitting = '';
			$this->currentAction = null;
		} else {
			echo "called end action when no action was current!";
		}
		return null;
	}

	function addHtml($html, $target){
		$act = new IWFAction('html', $target, $html);
		array_push($this->actions, $act);
		return $act;
	}

	function startHtml($target){
		return $this->startAction('html', $target);
	}

	function endHtml($close){
		if ($this->currentlyEmitting != 'html'){
			$this->error(-2163, 'iwf.endHtml() called, but iwf believes it should be emitting ' . $this->currentlyEmitting);
		} else {
			$this->endAction();
		}

		if ($close){
			$this->close();
		}
	}

	function error($code, $msg){
		if ($this->currentAction){
			$this->currentAction->error($code, $msg);
		} else {
			$this->addError($code, $msg);
		}
	}

	function endXml($finished){
		if ($this->currentlyEmitting != 'html'){
			$this->error(-2164, 'iwf.endXml() called, but iwf believes it should be emitting ' . $this->currentlyEmitting);
		} else {
			$this->endAction();
		}
		if (isset($finished) && $finished){
			$this->close();
		}
	}

	function addJavascript($js){
//		$act = new IWFAction('javascript', '', "<![CDATA[ $js ]]>" );
		$act = new IWFAction('javascript', '', "<!-- $js //-->" );
		array_push($this->actions, $act);
		return $act;
	}

	function popup($txt){
		return $this->addJavascript("alert('Server-side IWF popup:\\n$txt')");
	}

	function addXml($xml){
		$act = new IWFAction('xml', '', $xml);
		array_push($this->actions, $act);
		return $act;
	}

	function startXml(){
		return $this->startAction('xml', '');
	}

	function addError($code, $msg){
		$act = new IWFAction('','','');
		$act->error($code, $msg);
		array_push($this->actions, $act);
		return $act;
	}

	function _emit($emitEnd){
		if (!$this->startedEmitting){
			if ($this->iframeName != ''){
				echo "<html><head></head><body onload='javascript:window.parent._iwfIframeCallback(\"";
				echo $this->iframeName;
				echo "\", document.body.innerHTML);' >";
			}
			echo "<response";
			if ($this->debugging){
				echo " debugging='true'";
			}
			if ($this->prettyprint){
				echo " pretty='true'";
			}
			echo ">";
			$this->startedEmitting = true;
		}

		if ($this->currentAction){
			$this->endAction();
		}

		foreach($this->actions as $act){
			$act->emit(true, $this->prettyprint);
		}

		// empty the array, as we've emitted it and cannot go back anymore...
		$this->actions = array();

		if ($emitEnd){
			if ($this->prettyprint){
				echo "\n";
			}
			echo "</response>";
			if ($this->prettyprint){
				echo "\n";
			}
			if ($this->iframeName != ''){
				echo "</body></html>";
			}
		}
	}


	function close(){
		$this->_emit(true);
		$this->debugging = false;
	}

}

//function iwfXmlEncode($s){
//	return str_replace('"', '&quot;', str_replace("'", "&apos;", str_replace('>', '&gt;', str_replace('<', '&lt;', str_replace('&', '&amp;', $s)))));
//}

//function iwfXmlDecode($s){
//	return str_replace('&quot;', '"', str_replace('&apos;', "'", str_replace('&gt;', '>', str_replace('&lt;', '<', str_replace('&amp;', '&', $s)))));
//}

// create a singleton IWF object to use
$iwf = new IWFController();

$iwfIframeName = strtolower(@$_REQUEST['iwfIframeName']) . '';
$iwf->iframe($iwfIframeName);

$iwfPretty = strtolower(@$_REQUEST['iwfPretty']) == 'true';
$iwf->pretty($iwfPretty);

// debugging always sets prettyprint to true, so we must set debug last.
$iwfDebug = strtolower(@$_REQUEST['iwfDebug']) == 'true';
$iwf->debug($iwfDebug);



// load the default vars...
$iwfMode = strtolower(@$_REQUEST['iwfMode']);
$iwfId = @$_REQUEST['iwfId'];

$iwfTarget = @$_REQUEST['iwfTarget'];

//if (!isset($iwfTarget)){
//	$iwfTarget = 'iwfContent';
//}

?>

File: jas/webshare/iwf/ruby/iwf.rb
class IWFAction {

	def initialize(type, target, content)
		@type = type
		@target = target
		@content = content
		@emitted = false
	end

	def error(code, msg)
		@code = code
		@msg = msg
		@content = ''
		self
	end

	def emit_start(pretty_print)
		puts "\n\t" if pretty_print
		puts "<action type='{#@type}'"
		puts " target='{#@target}'" if @target
		puts " errorCode='{#@code}'" if @code
		puts " message='{#@message}'" if @msg
		puts ">"
		puts "\n\t" if pretty_print

		@emitted = true
		self
	end

	def emit_end(pretty_print)
		puts "\n\t" if pretty_print
		puts "</action>"
		self
	end

	def emit(end_emit, pretty_print)
		emit_start pretty_print
		puts @content
		emit_end if end_emit
		self
	end

}

class IWFController {

	def initialize()
		@actions = []
		@current_action = nil
		@currently_emitting = false
		@started_emitting = false
		@debugging = false
		@pretty_print = false
	end

	def debug(on)
		@debugging = on
		@pretty_print = true if on
	end

	def pretty(on)
		@pretty_print = on
	end

	def start_action(type, target)
		emit false
		@current_action = new IWFAction(type, target, '')
		@current_action.emit false, @pretty_print
		@currently_emitting = type
		return @current_action
	end


	def end_action
		if @current_action != nil {
			@current_action.emit true, @pretty_print
			@currently_emitting = ''
			@current_action = nil
		} else {
			puts "called end action when no action was current!"
		}
	end

	def add_html(html, target)
		action = new IWFAction('html', target, html)
		@actions.push action
		action
	end

	def start_html(target)
		start_action 'html', target
	end

	def end_html(close)
		unless @currently_emitting == 'html' {
			error -2163, "iwf.end_html called, but iwf believes it should be emitting {#@currently_emitting}"
		} else {
			end_action
		}
		self.close if close
	end

	def error(code, msg)
		if @current_action == nil
			add_error code, msg
		else
			@current_action.error(code, msg)
	end

	def end_xml(finished)
		unless @currently_emitting == 'xml' {
			error -2164, "iwf.end_xml called, but iwf believes it should be emitting {#@currently_emitting}"
		} else {
			end_action
		}

		close if finished

	end

	def add_js(js)
		add_javascript js
	end

#	alias add_js add_javascript
	def add_javascript(javascript)
		action = new IWFAction('javascript', '', "<![CDATA[ {#javascript} ]]>")
		@actions.push action
		action
	end

	def popup(text)
		add_javascript "alert('Server-side IWF popup:\n{#text}');"
	end

	def add_xml(xml)
		action = new IWFAction('xml', '', xml)
		@actions.push action
		action
	end

	def start_xml
		start_action xml, ''
	end

	def add_error(code, msg)
		action = new IWFAction('', '', '')
		action.error code, msg
		@actions.push action
		action
	end

	def emit(end_emit)
		unless @started_emitting {
			puts "<response"
			puts " debugging='true'" if @debugging
			puts " pretty='true'" if @pretty_print
			puts ">"
			@started_emitting = true
		}

		end_action unless @current_action == nil

		@actions.each { |act|,
			act.emit true, @pretty_print
		}

		actions = []

		if end_emit {
			puts '\n' if @pretty_print
			puts '</response>'
		}

	end

	def close
		emit true
		@debugging = false
	end

}


# create a singleton IWF object to use
# iwf = new IWFController()
# iwf.pretty(params[:iwfPretty] == 'true')

# debugging always sets prettyprint to true, so we must set debug last.
# iwf.debug(params[:iwfDebug] == 'true')

# load the default vars...
#iwf.mode = params[:iwfMode]
# iwf.id = params[:iwfId]
# iwf.target = params[:iwfTarget]
